### ISSUE TITLE
Worker support for BaseAudioContext

### ISSUE THREAD

**Issue by hoch** [WebAudio/web-audio-api-v2#16]
Currently the exposure of context constructors are not defined in the spec, but we can clarify by doing this:

```webidl
[Exposed=Window]
interface BaseAudioContext : EventTarget {
  ...
}
```

Furthermore, we can expose this to the WorkerGlobalScope (not Worklet):

```webidl
[Exposed=(Window, Worker)]
interface BaseAudioContext : EventTarget {
  ...
}
```

From the short investigation, I feel like we do not have anything tied to the Window or DOM in WebAudio. This is a relatively small spec change (non-breaking architectural change) but the advantage is tremendous in my opinion.

WDYT?

**Comment by hoch** [comment]
Action items from TPAC 2019:
- Talk to developers to collect the real-world use cases and prioritize.

**Comment by haywirez** [comment]
I haven't thought it completely through yet, but I'm wondering if future support for `SharedArrayBuffer`s changes the perspective on this issue as well.

My previous thinking was that most web audio apps will need to react to main thread UI / DOM interaction events (pointer, keyboard etc.) anyway. _Normally_, you'd want to use the listener callbacks to schedule events on the `AudioContext` straight from that thread, without the `postMessage` overhead. 

But with `SharedArrayBuffer`s  and a worker scope we could completely decouple the main UI thread by simply emitting timestamps + values into a `SharedArrayBuffer`. 

I guess this would necessitate another event loop of some kind inside the worker thread to pick up the events at its own pace (an `AudioWorklet`'s `process` is already kind of one). But even without `AudioWorkletNode`s in the graph, scheduling the parameters of other nodes from the worker could be useful. For performance management of the pick-up loop's frequency, you'd add a standard low-cost wake up / sleep `postMessage` trigger.

With this setup, the worker thread becomes a dedicated `AudioContext` event scheduler. This would also make sense considering that workers have support for networking APIs, access to IndexedDB etc., all of which can feed into the overall audio event schedule.  

The DOM/UI thread could then react separately, in a more precise manner (get the currently relevant timestamp values from a `SharedArrayBuffer` within a `requestAnimationFrame()` callback). It gets more and more complicated once you add separate 3D graphics threads (`OffscreenCanvas` rendering etc.) in to the mix... But again I'm not confident in the hidden costs involved, so not sure what would be the most sensible & elegant solution 😬

**Comment by padenot** [comment]
Virtual F2F:
- This has been requested quite a lot, and seems to be very useful for web apps that have a hard time managing the main thread load a scheduling
- This would work nicely in cunjonction with other APIs going off-main-thread such as Web MIDI.
- A solution needs to be found for `MediaStream` and `HTMLMediaElement`-related objects
- It's therefore not a small task, and we'll be dealing with this later than other issues
- I felt (and said) that the group making this (of course changeable) decision didn't have anybody working on an app that faces the problems outlined in this issue and so the decision is not very strong

**Comment by buildist** [comment]
This would be very helpful for my music sequencing web app (https://onlinesequencer.net/).

Right now clicking the play button starts a web worker, that worker periodically posts messages back to the main thread to play the next step. But the main thread is also doing a lot of graphical stuff at the same time, so sometimes there is lag, it would perform much better to do all of the audio processing in the worker. I know about AudioWorklet but it seems very tricky to use it for scheduling like this.

**Comment by roelvz** [comment]
> This would be very helpful for my music sequencing web app (https://onlinesequencer.net/).
> 
> Right now clicking the play button starts a web worker, that worker periodically posts messages back to the main thread to play the next step. But the main thread is also doing a lot of graphical stuff at the same time, so sometimes there is lag, it would perform much better to do all of the audio processing in the worker. I know about AudioWorklet but it seems very tricky to use it for scheduling like this.

Sounds familiar. We're retrieving encoded audio from an external source and it is being decoded on the main thread at the moment. This doesn't work well. I was also thinking about decoding on a web worker, and then posting messages to the main thread which can acces `AudioContext`. But indeed it would help a lot if the web worker could access the `AudioContext`.

**Comment by padenot** [comment]
TPAC 2020:

- It's well possible that having a first cut with no `MediaStream` is very useful
- We're asking the WebRTC Working Group to weigh in regarding transferable `MediaStream`s to extend this first cut

**Comment by AndreVanKammen** [comment]
I'm writing a synthesizer that uses webgl to calculate the samples. To use AudioWorklet I must post all the samples to the worklet thread, this gives overhead. The worklet has no access to OffscreenCanvas, but it would also hinder visualizers who need to access the same webglcontext. It would be nice if everything would run in a seperate worker. For now I changed the code back to use ScriptProcessorNode which gives me a lower latency.

https://shadersynth.com

**Comment by AndreVanKammen** [comment]
The support for SharedArrayBuffer is very low, I got my player to work on a snapdragon 845 on Android, but Android still does not support SharedArrayBuffer after years.
https://caniuse.com/sharedarraybuffer

**Comment by AshleyScirra** [comment]
I'd like to voice my support for using AudioContext in a web worker. It would be a huge simplification for our audio code in our web game engine [Construct 3](https://www.construct.net). Our engine is able to run entirely in a worker using OffscreenCanvas, but we lose access to the Web Audio API. So we have to have a very complex messaging system to essentially manage the entire Web Audio API. The messaging is also async, which adds further complications (especially regarding clocks, timing, and sample-accurate playback).

In practice we also need MediaStreams in workers, since we also integrate getUserMedia and MediaRecorder with Web Audio. Perhaps MediaStreams could be made transferable, so we can invoke those APIs in the DOM and just transfer the MediaStreams around.

**Comment by voxpelli** [comment]
I filed a Project Fugu request for this as well, some 1.5 years ago, I updated it with a link to eg. your comment @AshleyScirra: https://bugs.chromium.org/p/chromium/issues/detail?id=997514#c12

**Comment by hoch** [comment]
@voxpelli The issue 997514 is slightly different from what @AshleyScirra is asking.

**Comment by voxpelli** [comment]
@hoch Yes, there are lots of issues sprinkled across the web around audio in workers, they don't overlap a 100%, but still good to reference them all so that the overlap between them can be found.

The ServiceWorker case was the one I could come up with, the case @AshleyScirra come up with is a different one, though a also very valid one and with some similar challenges, in that one eg. can't depend on `HTMLMediaElement`

**Comment by padenot** [comment]
https://github.com/w3c/mediacapture-extensions/issues/16 is very helpful there, it removes the limitations outlined in https://github.com/WebAudio/web-audio-api-v2/issues/16#issuecomment-263952227. It's always possible to get a `MediaStreamTrack` from an `HTMLMediaElement`:

```js
var someAudioElement = document.querySelector("audio");
var theAudioTrack = someAudioElement.captureStream().getAudioTrack()[0];
someAudioElement.muted = true; // because it's probably going to be routed to the HTMLMediaElement
```
